contracts/ForkDelta.sol

```diff
@@ -347,4 +369,79 @@ contract ForkDelta {
     orderFills[msg.sender][hash] = amountGet;
     emit Cancel(tokenGet, amountGet, tokenGive, amountGive, expires, nonce, msg.sender, v, r, s);
   }
+  
+  ////////////////////////////////////////////////////////////////////////////////
+  // Contract Versioning / Migration
+  ////////////////////////////////////////////////////////////////////////////////
+  
+  /**
+  * User triggered function to migrate funds into a new contract to ease updates.
+  * Emits a FundsMigrated event.
+  * @param address Contract address of the new contract we are migrating funds to
+  * @param address[] Array of token addresses that we will be migrating to the new contract
+  */
+  function migrateFunds(address newContract, address[] tokens_) {
+  
+    require(newContract != address(0));
+    
+    ForkDelta newExchange = ForkDelta(successor);
+
+    // Move Ether into new exchange.
+    uint etherAmount = tokens[0][msg.sender];
+    if (etherAmount > 0) {
+      tokens[0][msg.sender] = 0;
+      newExchange.depositForUser.value(etherAmount)(msg.sender);
+    }
+
+    // Move Tokens into new exchange.
+    for (n = 0; n < tokens_.length; n++) {
+      address token = tokens_[n];
+      require(token != address(0)); // Ether is handled above.
+      uint tokenAmount = tokens[token][msg.sender];
+      
+      if (tokenAmount == 0) {
+        continue;
+      }
+      
+      require(Token(token).approve(newExchange, tokenAmount));
+      
+      tokens[token][msg.sender] = 0;
+      
+      newExchange.depositTokenForUser(token, tokenAmount, msg.sender);
+    }
+
+    FundsMigrated(msg.sender, newContract);
+  }
+  
+  /**
+  * This function handles deposits of Ether into the contract, but allows specification of a user.
+  * Note: This is intended for use in migration of funds.
+  * Note: With the payable modifier, this function accepts Ether.
+  */
+  function depositForUser(address user) public payable {
+    require(user != address(0));
+    require(msg.value > 0);
+    tokens[0][user] = tokens[0][user].add(msg.value);
+  }
+  
+  /**
+  * This function handles deposits of Ethereum based tokens into the contract, but allows specification of a user.
+  * Does not allow Ether.
+  * If token transfer fails, transaction is reverted and remaining gas is refunded.
+  * Note: This is intended for use in migration of funds.
+  * Note: With the payable modifier, this function accepts Ether.
+  * Note: Remember to call Token(address).approve(this, amount) or this contract will not be able to do the transfer on your behalf.
+  * @param token Ethereum contract address of the token or 0 for Ether
+  * @param amount uint of the amount of the token the user wishes to deposit
+  */
+  function depositTokenForUser(address token, uint amount, address user) public {
```

Comment:
+1 on removing the pseudo-check "require(caller.version() > 0);"
I can't spot a problem here, but we will need a number of test cases for this...